### DISCUSSION TITLE
make a new custom selector

### DISCUSSION THREAD

**Discussion by lgsmith** [GrossfieldLab/loos#97] [Developers]
I'd really like to define a new built-in selector (like 'backbone' or 'hydrogen'); actually I'd like to define several. I've poked around some but I'm not seeing how to do that (though I think that editing the contents of 'selectors' might allow me to change one that already exists). I can talk about the ones I want to create (in fact I'd like to, I've begun making one as a selection string and it was pretty involved...) but knowing how to make one in the abstract seems pretty helpful.

**Comment by agrossfield** [comment]
Summoning @tromo. I think you’d want to mimic HydrogenSelector.

**Comment by lgsmith** [comment]
That is what I was thinking too. I just don't see how to copy that so that I have some particular selection I can call from within a selection string. I can't seem to find where the selection strings are parsed, so that I can make a new name for one.

**Comment by tromo** [comment]
This is non-trivial, and you may find that the generated files from a modern flex/bison simply wont work.  What you will need to do is:

1. Add a selector to `Selectors.?pp`
2. Add a corresponding action to `KernalActions.?pp` (subclass `Action` ... see the hydrogen selector)
3. Modify `scanner.ll` to add a token for the new selector (again, see the hydrogen keyword)
4. Modify `grammer.yy` to recognize the new selector token as an expression and call the new action
5. Run scanner through `flex` and grammer through `bison`.
6. Rebuild and cross fingers, toes, light candles…

**Comment by agrossfield** [comment]
As a first pass, I suggest implementing it in python -- use a function with selection strings to test the strings and make sure everything works. Once you have all of that sorted, you can move on to trying to get it into the grammar.

**Comment by lgsmith** [comment]
Wait is the flex/bison not regenerated whenever you rebuild? Also, if not...and it doesn't work...shouldn't we fix that?

I'm trying to implement this as a selection string and it's taking forever. I suppose I can write something in python logic that does a bunch of testing on atom properties and connectivity...

**Reply by agrossfield** [reply]
Apparently flex/bison have changed in non-compatible ways, so we've been using the generated code (which doesn't change) for a long time. Some work may be required to get flex/bison working again. 

As for performance, @tromo can correct me but I don't think this will make it any faster. If I understand correctly, the selection strings get assembled by the parser into a function that says yes/no for each atom. If it's pre-compiled, the assembly part gets pre-done but the per-atom cost should be the same.

Unless I'm missing something, there shouldn't be any python/c++ logic in the protein hydrogen donor example you gave me:
```
('resname == "TYR" && name == "HO") || (next one) || (next one)
```

If you wanted to do h-bond donor/acceptors by chemical information, the Selection mechanism won't work, because by construction it considers atoms 1 at a time.

**Reply by lgsmith** [reply]
I was responding to something you said on slack: you suggested that I define it programmatically, by looking for hydrogens associated with oxygens (for example). 

If I can't add that kind of logic to a selector I suppose I'll just figure out whatever selection string I need to write, then add it as a selector later.

Alternatively, is there some way to teach the selection parser to do variable expansion so that we can define big complex strings like a macro that gets substituted into the strings before they're turned into selectors?

**Reply by tromo** [reply]
We also decided that it was unlikely that someone would want to muck around with the parser, so not generating the files each time made for a [slightly] simpler build.

As far as performance goes, the parser was never meant to be particularly performant.  Internally, it uses a simple stack-based machine and compiles a program representing the selection string (called a "kernel") which gets used by a selector (`KernelSelector()`) to actually do the selecting.  If your selection is invariant over a trajectory, then none of this matters since you select what atoms you want to work with at the start and then they all get updated through the magic of `pAtom`.  If you need to reselect per frame, then you're going to want to do it programmatically.  


FWIW, If you're using a shell script, you can always do something like,

`loos_command -s "( $(cat selection_one) ) &  ( $(cat selection_two) )" foo bar`

**Comment by lgsmith** [comment]
I know I can just save things in text files, but I kind of wanted to offer
people a way to select things without the headache, and I didn't want to
distribute data with my PR (though I know sometimes that's necessary). It
just seemed really graceful to be able to say `resid < 214 &&
polar_hydrogen`. We talk about wanting to have programs that read
conceptually, and I guess that seemed like the right concept to me.

**Comment by agrossfield** [comment]
My other issue with something like “polar_hydrogen” is that doing it *right* requires loos to learn chemistry, which I’ve strongly resisted having it do (my premise the programs should be either brilliant or stupid, because smart just gets you into trouble).

**Comment by lgsmith** [comment]
How else do we track hydrogen bonds across a large selection?

**Comment by agrossfield** [comment]
Right now, you have to identify the potential donors and acceptors yourself. You could guess programmatically (eg all hydrogens bound to something other than carbon are donors, but I’m sure there are lots of exceptions)

**Comment by tromo** [comment]
FWIW, I think the way I'd approach this is to write a function that picks out the polar hydrogens by whatever criteria you want.  You could add a command line flag that says only use polar hydrogens.  Your selection then is two stage: first use `selectAtoms()` and then call your function (if requested).  That way you can consider as many atoms as you want, or also make it sufficiently performant that you could call it every frame, if that's what you want to do.

**Comment by lgsmith** [comment]
I’d settled on this as well. That way if others want it they can call it.
